uart_stm32: fix a dead lock on poll_out if it's called in interrupt

jdascenzio opened the pull request:
If the function poll_out is called in an interrupt with a higher
priority than the usart priority, the function could lock forever. In
fact, the lock variable is never unset if the usart interrupt is no
longer called. So now, we don't use the lock variable if poll_out is
called in interrupt.

Fixes #40775

Signed-off-by: Julien D'ascenzio <julien.dascenzio@paratronic.fr>

erwango commented:
@jdascenzio Thanks, though I think there's a wider issue to fix, cf https://github.com/zephyrproject-rtos/zephyr/issues/40775

jdascenzio committed:
drivers/uart: stm32: fix dead lock on poll_out

A dead lock could happen if 2 threads with differents priorities use
poll_out. In fact, the lock data->tx_lock could be lock by a thread with
lower priority and then a thread with higher priority can't take the
lock. There was a race condition here:

/* Wait for TXE flag to be raised */
while (1) {
	if (atomic_cas(&data->tx_lock, 0, 1)) {
		/* !!!!!!!! RACE CONDITION !!!!!!!!!!!!!!
		if (LL_USART_IsActiveFlag_TXE(UartInstance)) {
			break;
		}
		atomic_set(&data->tx_lock, 0);
	}
}

To fix race condition, the interrupts are locked in poll_out.

Signed-off-by: Julien D'ascenzio <julien.dascenzio@paratronic.fr>

erwango reviewed (changes requested):
https://github.com/zephyrproject-rtos/zephyr/issues/40775 should be taken into account

jdascenzio commented:
I modified the patch to fix #40775. I hope this commit is the last to fix competitive calls to fifo_fill and uart_poll!

jdascenzio commented:
add lock to uart_stm32_irq_tx_enable and uart_stm32_irq_tx_disable to avoid race condition

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -525,21 +525,25 @@ static void uart_stm32_poll_out(const struct device *dev,
 	/* Wait for TXE flag to be raised */
 	while (1) {
-		if (atomic_cas(&data->tx_lock, 0, 1)) {
+		if (LL_USART_IsActiveFlag_TXE(UartInstance)) {
```
Please add a comment on what you're doing here since this is not obvious.
The commit message explain why previous solution was not working, 
but we also need to understand why current code is correct.
For this in line comments would be great

jdascenzio replied:
I have added comments which hopefully clarify the code

erwango reviewed (changes requested):
Some comments needed.
Meanwhile, I'm running broad tests and will keep you informed of the result

erwango reviewed (approved):
LGTM, and I confirm this fixes #40775

DariusBabrauskas reviewed (commented):
It looks Ok

DariusBabrauskas commented on drivers/serial/uart_stm32.c:
```diff
@@ -525,21 +525,32 @@ static void uart_stm32_poll_out(const struct device *dev,
 				break;
 			}
-			atomic_set(&data->tx_lock, 0);
+			irq_unlock(key);
```
if (!k_is_in_isr() && key == 0){			
			k_yield();//yield execution to another thread of the same or higher priority. Its better for software, but not output pretty (some symbols  from shell some from kprint)
		}

jdascenzio replied:
ok, k_yield could be interesting to reduce some cpu usage. But, I don't understand why you want check key?

DariusBabrauskas replied:
If key == 0 then  interrupts not disabled. So we can safe yield. I do not study  yield, when interrupts is disabled.

DariusBabrauskas reviewed (changes requested):
It seems to me, that need yield task if we not interrupt mode, and not disabled interrupts (k==0).

DariusBabrauskas reviewed (commented):
I do some self test on my system (stm32L4A6, zephyr 2.5) and noticed that I lost bytes on  terminal :( .
Exmp.  in switch interrupts kprint("ZzzzZ");  . On terminal output I often see zzzZ[0m  or zzzZJ... . So I lost first Z, but at and I see strange symbols.  Log messages lost too. hmm.

DariusBabrauskas reviewed (changes requested):
I noticed that irq_lock and irq_unlock don't have DSB() operation.  Possible her is cause of problems.  I can manualy add DSB for test purpose.

ycsin commented:
> I do some self test on my system (stm32L4A6, zephyr 2.5) and noticed that I lost bytes on terminal :( . Exmp. in switch interrupts kprint("ZzzzZ"); . On terminal output I often see zzzZ[0m or zzzZJ... . So I lost first Z, but at and I see strange symbols. Log messages lost too. hmm.

Do you have `CONFIG_PM` enabled in your build? It could be related to this issue #39220
The "strange bytes" (starts with `[` followed by `0m`/`J`/etc) are the control characters / color code for the shell.

DariusBabrauskas commented:
> > I do some self test on my system (stm32L4A6, zephyr 2.5) and noticed that I lost bytes on terminal :( . Exmp. in switch interrupts kprint("ZzzzZ"); . On terminal output I often see zzzZ[0m or zzzZJ... . So I lost first Z, but at and I see strange symbols. Log messages lost too. hmm.
> 
> Do you have `CONFIG_PM` enabled in your build? It could be related to this issue #39220 The "strange bytes" (starts with `[` followed by `0m`/`J`/etc) are the control characters / color code for the shell.

Yes you right. Than you.

DariusBabrauskas commented on drivers/serial/uart_stm32.c:
```diff
@@ -525,21 +525,36 @@ static void uart_stm32_poll_out(const struct device *dev,
 	while (1) {
-		if (atomic_cas(&data->tx_lock, 0, 1)) {
+		if (LL_USART_IsActiveFlag_TXE(UartInstance)) {
+			key = irq_lock();
```
__DSB();

DariusBabrauskas replied:
//My test show, that __DSB();   commands helped. No more lost first symbols!

DariusBabrauskas commented on drivers/serial/uart_stm32.c:
```diff
@@ -525,21 +525,36 @@ static void uart_stm32_poll_out(const struct device *dev,
 				break;
 			}
-			atomic_set(&data->tx_lock, 0);
+			irq_unlock(key);
```
__DSB();

DariusBabrauskas reviewed (changes requested):
It seems to me, that need change uart_stm32_fifo_fill function too. Protection from nested interrupts .
```
static int uart_stm32_fifo_fill(const struct device *dev,
				  const uint8_t *tx_data,
				  int size)
{
	USART_TypeDef *UartInstance = UART_STRUCT(dev);
	uint8_t num_tx = 0U;
	int key;

	if(LL_USART_IsActiveFlag_TXE(UartInstance))
	{
		key = irq_lock();//prevent nested interrupts or thread switch
		__DSB();
		while ((size - num_tx > 0) && //only max 2 bytes sent at time - no latency, because first byte in shift register second hold in TXD register
			   LL_USART_IsActiveFlag_TXE(UartInstance)) {
			/* TXE flag will be cleared with byte write to DR|RDR register */

			/* Send a character (8bit , parity none) */
			LL_USART_TransmitData8(UartInstance, tx_data[num_tx++]);
		}
		irq_unlock(key);
		__DSB();
	}

	return num_tx;
}
```

jdascenzio commented:
> It seems to me, that need change uart_stm32_fifo_fill function too. Protection from nested interrupts .
> 
> ```
> static int uart_stm32_fifo_fill(const struct device *dev,
> 				  const uint8_t *tx_data,
> 				  int size)
> {
> 	USART_TypeDef *UartInstance = UART_STRUCT(dev);
> 	uint8_t num_tx = 0U;
> 	int key;
> 
> 	if(LL_USART_IsActiveFlag_TXE(UartInstance))
> 	{
> 		key = irq_lock();//prevent nested interrupts or thread switch
> 		__DSB();
> 		while ((size - num_tx > 0) && //only max 2 bytes sent at time - no latency, because first byte in shift register second hold in TXD register
> 			   LL_USART_IsActiveFlag_TXE(UartInstance)) {
> 			/* TXE flag will be cleared with byte write to DR|RDR register */
> 
> 			/* Send a character (8bit , parity none) */
> 			LL_USART_TransmitData8(UartInstance, tx_data[num_tx++]);
> 		}
> 		irq_unlock(key);
> 		__DSB();
> 	}
> 
> 	return num_tx;
> }
> ```

You are right, it is necessary to protect here the reading of TXE flag.
However, I'm not familiar with memory/instruction barrier. But, if you say that there may be a problem with the order of the execution of the instructions (usage of DSB), after irq_[lock/unlock], then, the type of problem could happen every time these functions are used. It would then be perhaps more judicious to integrate this barrier directly in irq_[lock/unlock]....

cfriedt commented:
Is this ready to merge yet @jdascenzio ? Will you be adding the `__DSB()` calls that @DariusBabrauskas asked for?

erwango commented:
Setting DNM until the direction is clear.

jdascenzio commented:
I added a lock in fifo_fill function.
For me, irq_lock already has an instruction barrier with isb:
```c
static ALWAYS_INLINE unsigned int arch_irq_lock(void)
{
 # .......

#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
	unsigned int tmp;

	__asm__ volatile(
		"mov %1, %2;"
		"mrs %0, BASEPRI;"
		"msr BASEPRI_MAX, %1;"
		"isb;"
		: "=r"(key), "=r"(tmp)
		: "i"(_EXC_IRQ_DEFAULT_PRIO)
		: "memory");
```
@DariusBabrauskas, Don't you think that's enough?
The character that you were losing, was it not related to the forgotten lock in fifo_fill?

DariusBabrauskas commented:
> I added a lock in fifo_fill function. For me, irq_lock already has an instruction barrier with isb:
> 
> ```c
> static ALWAYS_INLINE unsigned int arch_irq_lock(void)
> {
>  # .......
> 
> #elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
> 	unsigned int tmp;
> 
> 	__asm__ volatile(
> 		"mov %1, %2;"
> 		"mrs %0, BASEPRI;"
> 		"msr BASEPRI_MAX, %1;"
> 		"isb;"
> 		: "=r"(key), "=r"(tmp)
> 		: "i"(_EXC_IRQ_DEFAULT_PRIO)
> 		: "memory");
> ```
> 
> @DariusBabrauskas, Don't you think that's enough? The character that you were losing, was it not related to the forgotten lock in fifo_fill?
Possible lost symbols was terminal log problem, which try hide escape sequences. I enabled binary log then I see all symbols. No lost with dsb or without (fifo_fill was edited).
So maybe you are right,- isb is enought.
It seem too me, that dsb must be after isb. I see dsb after isb in others systems. This operatios are different.

erwango commented:
Let's have a new round of review and test, considering the recent changes.

dcpleung commented on drivers/serial/uart_stm32.c:
```diff
@@ -615,6 +630,13 @@ static int uart_stm32_fifo_fill(const struct device *dev,
 	uint8_t num_tx = 0U;
+	int key;
+
+	if (!LL_USART_IsActiveFlag_TXE(UartInstance))
+		return 0;
```
Needs brackets even for a one line statement.

jdascenzio replied:
I had the brackets

dcpleung reviewed (changes requested):
Just a minor one about brackets.

Have this been tested in actual hardware, with different scenarios?

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -652,30 +677,44 @@ static int uart_stm32_fifo_read(const struct device *dev, uint8_t *rx_data,
+	int key;
 
-	LL_USART_DisableIT_TC(UartInstance);
+	key = irq_lock();
```
Is there a real reason to have a different behavior w/ and w/o PM ?

jdascenzio replied:
Yes, these different behaviors are used to deal with pm_constraint to prevent suspension (see e3f4907efe6438a4b28f17167c58a2e25865c295)

erwango commented:
@cfriedt This PR was tagged "Blocked: HW tests". It should not have been merged..
Hopefully the tests will be passed.

cfriedt commented:
> @cfriedt This PR was tagged "Blocked: HW tests". It should not have been merged..
> Hopefully the tests will be passed.

Apologies @erwango - best to request changes so that it does not make it past filters.

erwango commented:
@cfriedt Agree, but then, let's remove "Blocked: HW test", if it can be trusted.

erwango commented:
Ok, tests are passed. At least the in tree ones. I trust @jdascenzio to test the PM ones

jdascenzio commented:
No problems for me with CONFIG_PM. I don't lose any characters
